model/src/datetime/mod.rs

```diff
@@ -74,34 +71,41 @@ const ZERO_BYTE: u8 = b'0';
 /// [retrieving an ISO 8601 formatter]: Self::iso_8601
 /// [microseconds precision]: Self::as_micros
 /// [seconds precision]: Self::as_secs
+// We use a [`PrimitiveDateTime`] here since it does not store an offset, and
+// the API only operates in UTC. Additionally, it is four bytes smaller than an
+// [`OffsetDateTime`].
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub struct Timestamp(NonZeroU64);
+pub struct Timestamp(PrimitiveDateTime);
 
 impl Timestamp {
-    /// Create a non-zero timestamp from a Unix timestamp with microseconds
-    /// precision if the value is not zero.
-    pub const fn from_micros(unix_microseconds: u64) -> Option<Self> {
-        // `const`: Clippy suggests using `Option::map_or` which isn't const.
-        #[allow(clippy::option_if_let_else)]
-        if let Some(value) = NonZeroU64::new(unix_microseconds) {
-            Some(Self(value))
-        } else {
-            None
-        }
+    /// Create a timestamp from a Unix timestamp with microseconds precision.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`TimestampParseErrorType::Parsing`] error type if the parsing
+    /// failed.
+    ///
+    /// [`TimestampParseErrorType::Parsing`]: self::error::TimestampParseErrorType::Parsing
+    pub fn from_micros(unix_microseconds: i64) -> Result<Self, TimestampParseError> {
+        let nanoseconds = i128::from(unix_microseconds) * i128::from(NANOSECONDS_PER_MICROSECOND);
+
+        OffsetDateTime::from_unix_timestamp_nanos(nanoseconds)
+            .map(|offset| Self(PrimitiveDateTime::new(offset.date(), offset.time())))
+            .map_err(TimestampParseError::from_component_range)
     }
 
-    /// Create a non-zero timestamp from a Unix timestamp with seconds precision
-    /// if the value is not zero.
-    pub const fn from_secs(unix_seconds: u64) -> Option<Self> {
-        let micros = unix_seconds * MICROSECONDS_PER_SECOND;
-
-        // `const`: Clippy suggests using `Option::map_or` which isn't const.
-        #[allow(clippy::option_if_let_else)]
-        if let Some(value) = NonZeroU64::new(micros) {
-            Some(Self(value))
-        } else {
-            None
-        }
+    /// Create a timestamp from a Unix timestamp with seconds precision.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`TimestampParseErrorType::Parsing`] error type if the parsing
+    /// failed.
+    ///
```

Comment:
Could we keep a Range error type and the return that at this error instead of a parsing error?

Reply:
I can keep the Range error type but users can refer to the `source` so that we don't re-implement `time`s `Display` implementation. Does that sound alright?

Reply:
Yeah I think that is fine